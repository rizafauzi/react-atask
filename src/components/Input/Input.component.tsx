import { ChangeEvent, useCallback } from 'react';
import { styled } from 'styled-components';

import { ReactComponent as CloseIcon } from '@assets/icons/close.icon.svg';

interface InputProps {
  value?: string;
  className?: string;
  onChange?: (newValue: string) => void;
  onDiscard?: () => void;
}

const Input = ({ value, onChange, className, onDiscard }: InputProps) => {
  const handleChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      if (onChange) {
        onChange(event.target.value);
      }
    },
    [onChange]
  );

  const handleDisCard = useCallback(() => {
    if (onChange) {
      onChange('');
    }
    if (onDiscard) {
      onDiscard();
    }
  }, [onDiscard, onChange]);

  return (
    <InputWrapper>
      <input
        value={value}
        data-testid="input"
        className={className}
        onChange={handleChange}
        placeholder="Search User..."
      />
      {!!value && (
        <button className="icon" onClick={handleDisCard}>
          <CloseIcon height={20} width={20} />
        </button>
      )}
    </InputWrapper>
  );
};

const InputWrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  position: relative;

  .icon {
    right: 10px;
    position: absolute;

    &:hover {
      cursor: pointer;
    }
  }

  input {
    border: none;
    outline: none;
    padding: 6px 12px;
    border-radius: 6px;
    border: solid 0.5px #00000052;
    outline: solid 0px transparent;
    transition: all 0.2s ease-out;

    &:hover {
      outline: solid 3px #0000001e;
    }

    &:focus {
      border: solid 0.5px #000000b3;
    }
  }
`;

export default Input;
