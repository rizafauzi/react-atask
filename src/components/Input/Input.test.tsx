import { fireEvent, render, screen } from '@testing-library/react';

import Input from './Input.component';

const setup = () => {
  const utils = render(<Input />);
  const input: HTMLInputElement = screen.getByTestId('input');
  return {
    input,
    ...utils,
  };
};

test('should render Input component', () => {
  const value = 'Lorem Ipsum';
  const onChange = (string: string) => {
    console.info('test: ', string);
  };

  render(<Input value={value} onChange={onChange} />);
  const buttonElement = screen.getByTestId('input');
  expect(buttonElement).toBeInTheDocument();
});

test('It should change the value of the input', () => {
  const { input } = setup();
  fireEvent.change(input, { target: { value: 'rizafauzi' } });
  expect(input.value).toBe('rizafauzi');
});

test('It should allow value be deleted', () => {
  const { input } = setup();
  fireEvent.change(input, { target: { value: 'rizafauzi' } });
  expect(input.value).toBe('rizafauzi');
  fireEvent.change(input, { target: { value: '' } });
  expect(input.value).toBe('');
});

test('matches snapshot', () => {
  const value = 'Lorem Ipsum';
  const onChange = (string: string) => {
    console.info('test: ', string);
  };

  const input = render(<Input value={value} onChange={onChange} />);
  expect(input).toMatchSnapshot();
});
