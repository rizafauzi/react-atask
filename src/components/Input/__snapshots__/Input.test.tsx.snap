// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdvumH gJrIYL"
      >
        <input
          data-testid="input"
          placeholder="Search User..."
          value="Lorem Ipsum"
        />
        <button
          class="icon"
        >
          <svg
            fill="none"
            height="20"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              id="SVGRepo_bgCarrier"
              stroke-width="0"
            />
            <g
              id="SVGRepo_tracerCarrier"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <g
              id="SVGRepo_iconCarrier"
            >
              <path
                clip-rule="evenodd"
                d="M22 12C22 17.5228 17.5228 22 12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2C17.5228 2 22 6.47715 22 12ZM8.96963 8.96965C9.26252 8.67676 9.73739 8.67676 10.0303 8.96965L12 10.9393L13.9696 8.96967C14.2625 8.67678 14.7374 8.67678 15.0303 8.96967C15.3232 9.26256 15.3232 9.73744 15.0303 10.0303L13.0606 12L15.0303 13.9696C15.3232 14.2625 15.3232 14.7374 15.0303 15.0303C14.7374 15.3232 14.2625 15.3232 13.9696 15.0303L12 13.0607L10.0303 15.0303C9.73742 15.3232 9.26254 15.3232 8.96965 15.0303C8.67676 14.7374 8.67676 14.2625 8.96965 13.9697L10.9393 12L8.96963 10.0303C8.67673 9.73742 8.67673 9.26254 8.96963 8.96965Z"
                fill="#cfcfcf"
                fill-rule="evenodd"
              />
            </g>
          </svg>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdvumH gJrIYL"
    >
      <input
        data-testid="input"
        placeholder="Search User..."
        value="Lorem Ipsum"
      />
      <button
        class="icon"
      >
        <svg
          fill="none"
          height="20"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            id="SVGRepo_bgCarrier"
            stroke-width="0"
          />
          <g
            id="SVGRepo_tracerCarrier"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
          <g
            id="SVGRepo_iconCarrier"
          >
            <path
              clip-rule="evenodd"
              d="M22 12C22 17.5228 17.5228 22 12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2C17.5228 2 22 6.47715 22 12ZM8.96963 8.96965C9.26252 8.67676 9.73739 8.67676 10.0303 8.96965L12 10.9393L13.9696 8.96967C14.2625 8.67678 14.7374 8.67678 15.0303 8.96967C15.3232 9.26256 15.3232 9.73744 15.0303 10.0303L13.0606 12L15.0303 13.9696C15.3232 14.2625 15.3232 14.7374 15.0303 15.0303C14.7374 15.3232 14.2625 15.3232 13.9696 15.0303L12 13.0607L10.0303 15.0303C9.73742 15.3232 9.26254 15.3232 8.96965 15.0303C8.67676 14.7374 8.67676 14.2625 8.96965 13.9697L10.9393 12L8.96963 10.0303C8.67673 9.73742 8.67673 9.26254 8.96963 8.96965Z"
              fill="#cfcfcf"
              fill-rule="evenodd"
            />
          </g>
        </svg>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
